src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -920,15 +920,34 @@ void ShenandoahFreeSet::recycle_trash() {
   }
 
   // Relinquish the lock after this much time passed.
-  static constexpr jlong deadline_ns = 30000; // 30 us
+  jlong batch_start_time = 0;
+  jlong batch_end_time = os::javaTimeNanos();    // This value will be treated as the initial batch_start_time
+  // Process batches of regions until it has been 8 us since the last yield.
+  static constexpr jlong deadline_ns = 8000;     // 8 us
   size_t idx = 0;
   while (idx < count) {
-    os::naked_yield(); // Yield to allow allocators to take the lock
-    ShenandoahHeapLocker locker(_heap->lock());
-    const jlong deadline = os::javaTimeNanos() + deadline_ns;
-    while (idx < count && os::javaTimeNanos() < deadline) {
-      try_recycle_trashed(_trash_regions[idx++]);
+    if (idx > 0) {
+      os::naked_yield(); // Yield to allow allocators to take the lock, except on the first iteration
     }
+    ShenandoahHeapLocker locker(_heap->lock());
+    // Avoid another call to javaTimeNanos() if we already know time at which last batch ended
+    batch_start_time = batch_end_time;
+    const jlong deadline = batch_start_time + deadline_ns;
+    do {
+      // Measurements suggest it typically takes less than 200 ns on average to recycle_trash() for a single region.
+      // With an 8 us deadline and a 32-region batch, we find that we typically process 3 batches between yields.
+      //
+      // Note that deadline is enforced since the end of previous batch.  In the case that yield() or acquisition of heap lock
+      // takes a "long time", we will have less time to process regions, but we will always process at least one batch between
+      // yields.  Yielding more frequently when there is heavy contention for the heap lock or for CPU cores is considered the
+      // right thing to do.
+      const size_t REGION_STRIDE = 32;
```

Comment:
Maybe call this `REGIONS_PER_BATCH`? When I see `stride` I think how far the loop index moves on each iteration.

Reply:
Good suggestion.

Reply:
Thanks.  I've made this change.